File: aurora-core/src/main/java/net/gree/aurora/application/AuroraShardingConfigLoadServiceImpl.java
package net.gree.aurora.application;

import com.typesafe.config.Config;
import org.sisioh.dddbase.utils.Function1;
import org.sisioh.dddbase.utils.Try;

import java.io.File;

final class AuroraShardingConfigLoadServiceImpl implements AuroraShardingConfigLoadService {

    private final ShardingConfigModelsParser parser = new ShardingConfigModelsParser();

    private final Function1<ShardingConfigModelsParser.ParseResult, Repositories> f = new Function1<ShardingConfigModelsParser.ParseResult, Repositories>() {
        @Override
        public Repositories apply(ShardingConfigModelsParser.ParseResult value) {
            return new RepositoriesImpl(value.dataSourceRepository, value.shardingConfigRepository);
        }
    };

    @Override
    public Try<Repositories> loadFromConfigFile(File configFile) {
        return parser.parse(configFile).map(f);
    }

    @Override
    public Try<Repositories> loadFromConfig(Config config) {
        return parser.parse(config).map(f);
    }

}

File: aurora-scala/src/main/scala/net/gree/aurora/scala/application/AuroraShardingConfigLoadServiceImpl.scala
package net.gree.aurora.scala.application

import java.io.File
import net.gree.aurora.application.{AuroraShardingConfigLoadService => JAuroraShardingConfigLoadService}
import net.gree.aurora.application.{AuroraTableNameService => JAuroraTableNameService}
import net.gree.aurora.application.{Repositories => JRepositories}
import net.gree.aurora.domain.tablename.{TableName => JTableName}
import org.sisioh.config.Configuration
import org.sisioh.dddbase.utils.{Try => JTry}
import scala.util.{Failure, Success, Try}

private[application]
class AuroraShardingConfigLoadServiceImpl(val underlying: JAuroraShardingConfigLoadService) extends AuroraShardingConfigLoadService {

  def loadFromConfigFile(configFile: File): Try[Repositories] = {
    val result = underlying.loadFromConfigFile(configFile)
    result match {
      case success: JTry.Success[_] =>
        val _result = success.get().asInstanceOf[JRepositories]
        Success(RepositoriesImpl(_result))
      case failure: JTry.Failure[_] =>
        Failure(failure.getCause)
    }
  }

  def loadFromConfig(config: Configuration): Try[Repositories] = {
    val result = underlying.loadFromConfig(config.underlying)
    result match {
      case success: JTry.Success[_] =>
        val _result = success.get().asInstanceOf[JRepositories]
        Success(RepositoriesImpl(_result))
      case failure: JTry.Failure[_] =>
        Failure(failure.getCause)
    }
  }

}
